google/gax/retry.py

```diff
@@ -0,0 +1,157 @@
+            except errors.RetryError as exception:
+                raise exception
+            except Exception as exception:  # pylint: disable=broad-except
+                code = config.exc_to_code(exception)
+                if code not in retry_options.retry_codes:
+                    raise errors.RetryError(
+                        'Exception occurred in retry method that was not'
+                        ' classified as transient', exception)
+
+                to_sleep = random.uniform(0, delay)
+                time.sleep(to_sleep / _MILLIS_PER_SECOND)
+                delay = min(delay * delay_mult, max_delay)
+
+    return inner
+
+
+def add_timeout_arg(a_func, timeout, **kwargs):
+    """Updates a_func so that it gets called with the timeout as its final arg.
+
+    This converts a callable, a_func, into another callable with an additional
+    positional arg.
+
+    Args:
+      a_func (callable): a callable to be updated
+      timeout (int): to be added to the original callable as it final positional
+        arg.
+
+
+    Returns:
+      callable: the original callable updated to the timeout arg
+    """
+
+    def inner(*args):
+        """Updates args with the timeout."""
+        updated_args = args + (timeout,)
+        return a_func(*updated_args, **kwargs)
+
+    return inner
+
+
+def _retryable_with_timeout(a_func, retry_options, **kwargs):
+    timeout_mult = retry_options.backoff_settings.rpc_timeout_multiplier
+    max_timeout = (retry_options.backoff_settings.max_rpc_timeout_millis /
+                   _MILLIS_PER_SECOND)
+    total_timeout = (retry_options.backoff_settings.total_timeout_millis /
+                     _MILLIS_PER_SECOND)
+
+    def inner(*args):
+        """Equivalent to ``a_func``, but retries with a timeout upon transient
+           failure.
+        """
+        timeout = (retry_options.backoff_settings.initial_rpc_timeout_millis /
+                   _MILLIS_PER_SECOND)
+        exc = errors.RetryError('Retry total timeout exceeded before any'
+                                'response was received')
+        now = time.time()
+        deadline = now + total_timeout
+        while now < deadline:
+            try:
+                to_call = add_timeout_arg(a_func, timeout, **kwargs)
+                return to_call(*args)
+            except Exception as exception:  # pylint: disable=broad-except
+                code = config.exc_to_code(exception)
+                if code not in retry_options.retry_codes:
+                    raise exception
+
+                # pylint: disable=redefined-variable-type
+                exc = errors.RetryError(
+                    'Retry total timeout exceeded with exception', exception)
+                now = time.time()
+                timeout = min(
+                    timeout * timeout_mult, max_timeout, deadline - now)
+
+        raise exc
+
+    return _retryable_without_timeout(inner, retry_options, **kwargs)
```

Comment:
This looks to me like it retries `inner`, and `inner` itself retries with no delay; you effective have nested retry `while` loops. So if I understand right, it'll do something like:

```
call1: timeout 10, delay 1
call2: timeout 20, delay 1
call3: timeout 40, delay 1
call4: timeout 10, delay 1.5
call5: timeout 20, delay 1.5
call6: timeout 40, delay 1.5
call7: timeout 10, delay 2.25
...
```

instead of

```
call1: timeout 10, delay 1
call2: timeout 20, delay 1.5
call3: timeout 40, delay 2.25
...
```

I think this might be more easily done with a single retryable function essentially identical the one previously in `api_callable` with some conditionals only to update the timeout if applicable. Sort of like (not real Python):

```
def retryable(...):
  def inner(...):
    ...
    timeout = {timeout setting} if {timeout setting} else None
    now = time.time()
    deadline = now + {total_timeout} if {total_timeout} else None
    while deadline is None or now < deadline
      try:
        to_call = a_func
        if timeout:
         to_call = add_timeout_arg(to_call, ...)
      except:
        [update delay]
        if timeout:
          [update timeout]
     ...
```

Will something like that work here?